### PR TITLE
feat: refactor ws impl and add test framework

### PR DESCRIPTION
**Pull request by jensneuse**

### CODE REVIEW

File: router/core/websocket.go [hunk 1]
```diff
@@ -158,76 +128,215 @@ func (c *wsConnectionWrapper) Close() error {
...
-			CheckOrigin: func(_ *http.Request) bool {
-				// Allow any origin to subscribe via WS
+		http.Error(w, http.StatusText(statusCode), statusCode)
+		return
+	}
+	r = validatedReq
+	var (
+		subProtocol string
+	)
+	upgrader := ws.HTTPUpgrader{
```
Comment (StarpTech) [inline router/core/websocket.go:169]: Can we reuse it across requests?

File: router/core/websocket.go [hunk 2]
```diff
@@ -158,76 +128,215 @@ func (c *wsConnectionWrapper) Close() error {
...
+				h.logger.Warn("handling websocket connection", zap.Error(err))
+				return
+			}
 		}
-		conn := newWSConnectionWrapper(h.ctx, c)
-		protocol, err := wsproto.NewProtocol(c.Subprotocol(), conn)
+	}
+}
+
+func (h *WebsocketHandler) isReadTimeout(err error) bool {
```
Comment (StarpTech) [inline router/core/websocket.go:253]: I'd make it a pure function. No data is shared. Maybe we can even reuse it somewhere else.

File: router/core/flushwriter.go [hunk 3]
```diff
@@ -122,6 +121,18 @@ func GetFlushWriter(ctx *resolve.Context, variables []byte, r *http.Request, w h
 		setSubscriptionHeaders(w)
 	}
 
+	w.WriteHeader(http.StatusOK)
```
Comment (StarpTech) [inline router/core/flushwriter.go:124]: I'd a comment. This is important to respond early to the upgrade right?

File: router/config/config.go [hunk 4]
```diff
@@ -150,21 +150,26 @@ type RequestHeaderRule struct {
...
 type EngineDebugConfiguration struct {
-	PrintOperationTransformations bool `envconfig:"ENGINE_DEBUG_PRINT_OPERATION_TRANSFORMATIONS"`
-	PrintOperationEnableASTRefs   bool `envconfig:"ENGINE_DEBUG_PRINT_OPERATION_ENABLE_AST_REFS"`
-	PrintPlanningPaths            bool `envconfig:"ENGINE_DEBUG_PRINT_PLANNING_PATHS"`
-	PrintQueryPlans               bool `envconfig:"ENGINE_DEBUG_PRINT_QUERY_PLANS"`
-	PrintNodeSuggestions          bool `envconfig:"ENGINE_DEBUG_PRINT_NODE_SUGGESTIONS"`
-	ConfigurationVisitor          bool `envconfig:"ENGINE_DEBUG_CONFIGURATION_VISITOR"`
-	PlanningVisitor               bool `envconfig:"ENGINE_DEBUG_PLANNING_VISITOR"`
-	DatasourceVisitor             bool `envconfig:"ENGINE_DEBUG_DATASOURCE_VISITOR"`
+	PrintOperationTransformations bool `default:"false" envconfig:"ENGINE_DEBUG_PRINT_OPERATION_TRANSFORMATIONS" yaml:"print_operation_transformations"`
```
Comment (StarpTech) [inline router/config/config.go:153]: You made them configurable in yaml. Please update the docs.

File: router/config/config.go [hunk 5]
```diff
@@ -150,21 +150,26 @@ type RequestHeaderRule struct {
...
+	ReportWebSocketConnections    bool `default:"false" envconfig:"ENGINE_DEBUG_REPORT_WEBSOCKET_CONNECTIONS" yaml:"report_websocket_connections"`
+	ReportMemoryUsage             bool `default:"false" envconfig:"ENGINE_DEBUG_REPORT_MEMORY_USAGE" yaml:"report_memory_usage"`
 }
 
 type EngineExecutionConfiguration struct {
-	Debug                                  EngineDebugConfiguration
-	EnableSingleFlight                     bool `default:"true" envconfig:"ENGINE_ENABLE_SINGLE_FLIGHT"`
-	EnableRequestTracing                   bool `default:"true" envconfig:"ENGINE_ENABLE_REQUEST_TRACING"`
-	EnableExecutionPlanCacheResponseHeader bool `default:"false" envconfig:"ENGINE_ENABLE_EXECUTION_PLAN_CACHE_RESPONSE_HEADER"`
+	Debug                                  EngineDebugConfiguration `yaml:"debug"`
```
Comment (StarpTech) [inline router/config/config.go:166]: Here too

File: router/config.yaml [hunk 6]
```diff
@@ -1,4 +1,96 @@
 # config.yaml
 
 # See pkg/config/config.go for the full list of configuration options.
-# Please keep this file empty if you don't want to override the defaults.
\ No newline at end of file
+# Please keep this file empty if you don't want to override the defaults.
```
Comment (StarpTech) [inline router/config.yaml:4]: Please keep it empty.

File: router/__schemas/compose.sh [hunk 7]
```diff
@@ -0,0 +1,10 @@
+#!/bin/bash
```
Comment (StarpTech) [inline router/__schemas/compose.sh:1]: Those files are not required for the router. Can we move them to router_tests?

File: demo/cmd/subscriptiontest/run.sh [hunk 8]
```diff
@@ -0,0 +1,6 @@
+#!/usr/bin/env sh
+export CGO_ENABLED=0
+export GOOS=linux
+export GOARCH=amd64
+go build --tags "static netgo" -o subscriptiontest main.go
+docker run --rm -it --platform linux/amd64 -l subscriptiontest -v $(pwd)/subscriptiontest:/subscriptiontest alpine /subscriptiontest -instances=100 -host=host.docker.internal
```
Comment (StarpTech) [inline demo/cmd/subscriptiontest/run.sh:6]: Could you document that a bit in the readme?

File: router-tests/events_test.go [hunk 9]
```diff
@@ -1,113 +1,401 @@
...
-		subscriptionID, err := client.Subscribe(&subscription, nil, func(dataValue []byte, errValue error) error {
-			require.NoError(t, errValue)
-			assert.JSONEq(t, `{"employeeUpdated":{"id":3,"details":{"forename":"Stefan","surname":"Avram"}}}`, string(dataValue))
-			triggers.Inc()
-			return nil
+	})
+
+	// times out on github actions
+
+	/*t.Run("subscribe async epoll/kqueue disabled", func(t *testing.T) {
```
Comment (StarpTech) [inline router-tests/events_test.go:86]: Lets use t.skip here.

File: router-tests/authentication_test.go
Comment (StarpTech) [inline router-tests/authentication_test.go:1]: Where you able to fix the races for websockets? Your PR does not test with `-race` enabled in the ci. You have to modify the Makefile in router_tests. There is a comment by me.

File: router-tests/events_test.go [hunk 10]
```diff
@@ -1,113 +1,401 @@
...
-			err := msg.Respond([]byte(`{"id": 3, "__typename": "Employee"}`))
+			sub, err := xEnv.NC.Subscribe("getEmployee.3", func(msg *nats.Msg) {
+				err := msg.Respond([]byte(`{"id": 3, "__typename": "Employee"}`))
+				require.NoError(t, err)
+			})
 			require.NoError(t, err)
+
+			t.Cleanup(func() {
+				err := sub.Unsubscribe()
+				assert.NoError(t, err)
```
Comment (StarpTech) [inline router-tests/events_test.go:354]: I'd always use `require.noError` because `assert` is handled optionally.

**Review by StarpTech — CHANGES_REQUESTED** [review]

File: router-tests/module/module_test.go
Comment (StarpTech) [inline router-tests/module/module_test.go:1]: Why did you remove the test?

File: router-tests/module-jwt/module_test.go
Comment (StarpTech) [inline router-tests/module-jwt/module_test.go:1]: Why did you remove the test?

**Review by StarpTech — CHANGES_REQUESTED** [review]

File: router/core/flushwriter.go (hunk 3 above)
Comment (jensneuse) [reply router/core/flushwriter.go:124]: delete

File: router/core/flushwriter.go [hunk 11]
```diff
@@ -122,6 +121,18 @@ func GetFlushWriter(ctx *resolve.Context, variables []byte, r *http.Request, w h
...
+
+	/*// hijack the connection
```
[the remark this answers is not in the document]
Comment (jensneuse) [reply router/core/flushwriter.go:126]: yes

File: router/core/websocket.go (hunk 1 above)
Comment (jensneuse) [reply router/core/websocket.go:169]: yes, I think so but please test it

File: router/core/websocket.go (hunk 2 above)
Comment (jensneuse) [reply router/core/websocket.go:253]: yes

File: router/__schemas/compose.sh (hunk 7 above)
Comment (StarpTech) [reply router/__schemas/compose.sh:1]: I kept them and document it.

File: router-tests/authentication_test.go
Comment (StarpTech) [reply router-tests/authentication_test.go:1]: Works. Tested locally.

File: router/core/websocket.go (hunk 1 above)
Comment (StarpTech) [reply router/core/websocket.go:169]: I skipped this optimization because there is no easy way to access the protocol afterwards.

**Comment by devsergiy** [comment]
@StarpTech please do not merge this when it's done

I will need to do a hotfix on the engine before merging this

**Review by StarpTech — APPROVED** [review]
LGTM

**Review by StarpTech — APPROVED** [review]
LGTM

**Pull request merged**